netlib/http2/protocol.py

```diff
@@ -48,16 +49,19 @@ def check_alpn(self):
                 "HTTP2Protocol can not handle unknown ALP: %s" % alp)
         return True
 
-    def _receive_settings(self):
-        frm = frame.Frame.from_file(self.tcp_handler.rfile, self)
-        assert isinstance(frm, frame.SettingsFrame)
-        self._apply_settings(frm.settings)
+    def _receive_settings(self, hide=False):
+        while True:
+            frm = self.read_frame(hide)
+            if isinstance(frm, frame.SettingsFrame):
+                break
```

Comment:
I'm merging, but I'd like the API to give us some way to access discarded frames. In the mitmproxy UIs, we'll probably have something to toggle between a frame view and an HTTP view, and we'll need the frame view to be complete. If this means keeping state on something stored on the connection, we should bake that into this low-level API.